### PR TITLE
test: refactor withdraw tests to cover all edge cases

### PR DESCRIPTION
**Pull request by smol-ninja**
### Changelog
- Closes https://github.com/sablier-labs/flow/issues/238.
- Renames `updateSnapshotTimeAndWarp` function to `takeSnapshot` and removes `warp` logic from it.
- Renames `WARP_ONE_MONTH` constant to `ONE_MONTH_SINCE_START`. 
    - The variable name `WARP_ONE_MONTH` goes well when called in `vm.warp` but not when used as a variable. For example in `uint40 exptectedDepletionTime = WARP_ONE_MONTH + uint40(solvencyPeriod + 1)`.

### CODE REVIEW

File: tests/integration/concrete/withdraw/withdraw.tree [hunk 1]
```diff
@@ -14,36 +14,45 @@ Withdraw_Integration_Concrete_Test
             ├── when withdrawal address not owner
             │  ├── when caller sender
             │  │  └── it should revert
-            │  ├── when caller unknown
-            │  │  └── it should revert
-            │  └── when caller recipient
-            │     └── it should withdraw
```
Comment (andreivladbrg) [inline tests/integration/concrete/withdraw/withdraw.tree:20]: why was this branch removed from here?

there is no test for the case when:

`msg.sender == recipient && to != recipient`

if it is due to bulloak, i think we can use the "caller owner" terminology

File: tests/utils/Vars.sol [hunk 2]
```diff
@@ -11,6 +11,7 @@ struct Vars {
     // previous values.
     uint256 previousAggregateAmount;
     uint256 previousOngoingDebtScaled;
+    uint256 previousSnapshotDebtScaled;
```
Comment (andreivladbrg) [inline tests/utils/Vars.sol:14]: why was this added? 
it is not used anywhere

File: tests/integration/concrete/withdraw/withdraw.t.sol [hunk 3]
```diff
@@ -186,32 +256,37 @@ contract Withdraw_Integration_Concrete_Test is Shared_Integration_Concrete_Test
...
+        whenCallerRecipient
+        givenBalanceExceedsTotalDebt
+        whenAmountLessThanTotalDebt
+        whenAmountGreaterThanSnapshotDebt
     {
-        // Go back to the starting point.
-        vm.warp({ newTimestamp: OCT_1_2024 });
-
-        resetPrank({ msgSender: users.sender });
+        // Turn on the protocol fee for usdc.
```
Comment (andreivladbrg) [inline tests/integration/concrete/withdraw/withdraw.t.sol:264]: should we capitalize to `USDC`?

File: tests/integration/concrete/withdraw/withdraw.t.sol [hunk 4]
```diff
@@ -251,87 +333,88 @@ contract Withdraw_Integration_Concrete_Test is Shared_Integration_Concrete_Test
...
+        vars.expectedProtocolRevenue = flow.protocolRevenue(vars.token) + vars.expectedProtocolFeeAmount;
+        vars.expectedWithdrawAmount = withdrawAmount - vars.expectedProtocolFeeAmount;
 
         // It should emit 1 {Transfer}, 1 {WithdrawFromFlowStream} and 1 {MetadataUpdated} events.
         vm.expectEmit({ emitter: address(vars.token) });
-        emit IERC20.Transfer({ from: address(flow), to: to, value: withdrawAmount - protocolFeeAmount });
+        emit IERC20.Transfer({ from: address(flow), to: to, value: vars.expectedWithdrawAmount });
 
-        vm.expectEmit({ emitter: address(flow) });
+        // vm.expectEmit({ emitter: address(flow) });
```
Comment (andreivladbrg) [inline tests/integration/concrete/withdraw/withdraw.t.sol]: i think typo
same below

File: tests/integration/concrete/withdraw/withdraw.t.sol [hunk 5]
```diff
@@ -251,87 +333,88 @@ contract Withdraw_Integration_Concrete_Test is Shared_Integration_Concrete_Test
...
-            protocolFeeAmount: 0,
-            withdrawAmount: WITHDRAW_AMOUNT_6D
+            withdrawAmount: amountToWithdraw
         });
+
+        // It should set snapshot debt to difference between total debt and amount withdrawn.
+        assertEq(
+            flow.getSnapshotDebtScaled(defaultStreamId),
+            getScaledAmount(totalDebt - amountToWithdraw, 6),
+            "snapshot debt"
```
Comment (andreivladbrg) [inline tests/integration/concrete/withdraw/withdraw.t.sol:357]: why not move this assert in `_test_Withdraw` since, it is suppose to be in 18 decimals

and keep only the `snapshot time` assert to individual tests

File: tests/integration/concrete/withdraw/withdraw.t.sol [hunk 6]
```diff
@@ -251,87 +333,88 @@ contract Withdraw_Integration_Concrete_Test is Shared_Integration_Concrete_Test
...
+        );
+        // It should update snapshot time to current time
+        assertEq(flow.getSnapshotTime(defaultStreamId), getBlockTimestamp(), "snapshot debt");
```
Comment (andreivladbrg) [inline tests/integration/concrete/withdraw/withdraw.t.sol]: incorrect error string reason

should be "snapshot time" - similar across the file

**Review by andreivladbrg — COMMENTED** [review]
great work, left some comments bellow:

File: tests/integration/concrete/withdraw/withdraw.t.sol (hunk 6 above)
Comment (andreivladbrg) [reply tests/integration/concrete/withdraw/withdraw.t.sol]: bf259891f327b49acc808599ba0be915e54e50eb

File: tests/integration/concrete/withdraw/withdraw.t.sol (hunk 4 above)
Comment (andreivladbrg) [reply tests/integration/concrete/withdraw/withdraw.t.sol]: bf259891f327b49acc808599ba0be915e54e50eb

File: tests/utils/Vars.sol (hunk 2 above)
Comment (smol-ninja) [reply tests/utils/Vars.sol:14]: Good catch. I added it to use but ended up not using it.

File: tests/integration/concrete/withdraw/withdraw.t.sol (hunk 5 above)
Comment (smol-ninja) [reply tests/integration/concrete/withdraw/withdraw.t.sol:357]: There are three places: `test_WhenAmountEqualsTotalDebt`, `test_WhenAmountLessThanSnapshotDebt` and `test_WhenAmountEqualsSnapshotDebt` where there are different assertions for snapshot debt. Thats why I put them into individual tests.

File: tests/integration/concrete/withdraw/withdraw.tree (hunk 1 above)
Comment (smol-ninja) [reply tests/integration/concrete/withdraw/withdraw.tree:20]: I made the following changes:

1. When withdrawal address is not owner: check for sender, unknown and recipient callers because it should revert for the first two.
2. When withdrawal address is owner: use only recipient as the caller. Caller is no more relevant in this case. For completeness, we have `testFuzz_Withdraw` where we are fuzzing caller as well.
    - I could run tests for all three callers using a modifier but then it makes the code ugly and difficult to read since I will have to go back and forth in time to reset stream IDs for each caller.

File: tests/integration/Integration.t.sol [hunk 7]
```diff
@@ -62,18 +62,15 @@ abstract contract Integration_Test is Base_Test {
         deposit(streamId, depositAmount);
     }
 
-    /// @dev Update the snapshot using `adjustRatePerSecond` and then warp block timestamp to it.
-    function updateSnapshotTimeAndWarp(uint256 streamId) internal {
+    /// @dev Take the snapshot using `adjustRatePerSecond`.
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply tests/integration/Integration.t.sol]: https://github.com/sablier-labs/flow/pull/368/commits/f20dd8434393aa691c8a1663d09b1b9c318e022b: Also renamed function name to `updateSnapshot`.

File: tests/integration/concrete/withdraw/withdraw.t.sol (hunk 3 above)
Comment (smol-ninja) [reply tests/integration/concrete/withdraw/withdraw.t.sol]: https://github.com/sablier-labs/flow/pull/368/commits/f20dd8434393aa691c8a1663d09b1b9c318e022b

File: tests/integration/concrete/withdraw/withdraw.tree (hunk 1 above)
Comment (andreivladbrg) [reply tests/integration/concrete/withdraw/withdraw.tree:20]: > Caller is no more relevant in this case. 
hmm, i am not sure why say this i believe we should test each individual case here as well

> For completeness, we have testFuzz_Withdraw where we are fuzzing caller as well

IMO fuzz tests are meant to catch cases we couldn't think of in unit or integration tests

Comment (andreivladbrg) [reply tests/integration/concrete/withdraw/withdraw.tree:20]: btw, if we keep the current version we could remove the `to` param in the `_test_Withdraw` function as `users.recipient` is passed on each test

File: tests/integration/concrete/withdraw/withdraw.t.sol [hunk 8]
```diff
@@ -232,16 +299,20 @@ contract Withdraw_Integration_Concrete_Test is Shared_Integration_Concrete_Test
...
-            protocolFeeAmount: 0,
-            withdrawAmount: WITHDRAW_AMOUNT_18D
-        });
+            withdrawAmount: WITHDRAW_AMOUNT_18D // < total debt and > snapshot debt
+         });
+
+        // It should set snapshot debt to difference between total debt and amount withdrawn.
+        assertEq(flow.getSnapshotDebtScaled(streamId), ONE_MONTH_DEBT_18D - WITHDRAW_AMOUNT_18D, "snapshot debt");
+        // It should update snapshot time to current time
+        assertEq(flow.getSnapshotTime(defaultStreamId), getBlockTimestamp(), "snapshot time");
```
Comment (andreivladbrg) [inline tests/integration/concrete/withdraw/withdraw.t.sol:315]: here we are testing the same stream:

```solidity
//                                     ↓ notice the ID
assertEq(flow.getSnapshotDebtScaled(streamId)
//                                     ↓ notice the ID
assertEq(flow.getSnapshotTime(defaultStreamId)
```

**Review by andreivladbrg — COMMENTED** [review]
just noticed one more thing

File: tests/integration/concrete/withdraw/withdraw.t.sol (hunk 5 above)
Comment (andreivladbrg) [reply tests/integration/concrete/withdraw/withdraw.t.sol:345]: you are correct, we should keep this way

File: tests/integration/concrete/withdraw/withdraw.tree (hunk 1 above)
Comment (smol-ninja) [reply tests/integration/concrete/withdraw/withdraw.tree:20]: > why say this i believe we should test each individual case here as well

I think, in concrete tests, whether a case should be tested or not should depend on if it directly influences the outcome or not. So when withdrawal address is not owner, value of caller impacts the result whereas, when its not owner, value of caller does not impact the result. 

Consider a simple example. Suppose there’s a deposit function, `f(amount)`, that accepts deposits of any amount including 0. There, there is another function, `g(amount)`, which accepts deposits of any amount but not 0. 

Now in the case of `f(amount)`, will we be testing how the function behaves when amount = 0, in concrete tests? I’d say probably not (we don't check for all values for which tests work). However, for `g(amount)` we would absolutely check if it reverts when amount = 0. Wdyt? 

PS: These are just some thoughts I’m brainstorming so feel free to share any opinions or feedback you have.

---------------- 

> if we keep the current version we could remove the to param in the _test_Withdraw function as users.recipient is passed on each test

`test_WhenCallerRecipient` calls `_test_Withdraw` with `users.eve` as the recipient.

File: tests/integration/concrete/withdraw/withdraw.t.sol (hunk 8 above)
Comment (smol-ninja) [reply tests/integration/concrete/withdraw/withdraw.t.sol:315]: yes it should be `streamId`

Comment (smol-ninja) [reply tests/integration/concrete/withdraw/withdraw.t.sol]: https://github.com/sablier-labs/flow/pull/368/commits/3f9bac38ddcd3e9c7e6ee3cf6cd74d10a85df7a9

File: tests/integration/concrete/withdraw/withdraw.tree (hunk 1 above)
Comment (andreivladbrg) [reply tests/integration/concrete/withdraw/withdraw.tree:20]: > Now in the case of f(amount), will we be testing how the function behaves when amount = 0, in concrete tests?

i don't think it is the same. a better comparison would have been:
1. `g(amount)` does not accept `amount == 0`
2. `f(amount)` accepts `amount == 0` only a certain criteria is met
   - and we test the case when that criteria is met

*(similar to our example)*
which IMO we should test.

---

> test_WhenCallerRecipient calls _test_Withdraw with users.eve as the recipient

right, missed it

Comment (smol-ninja) [reply tests/integration/concrete/withdraw/withdraw.tree:20]: OK I an convinced. I will run these tests for all three.

Comment (andreivladbrg) [reply tests/integration/concrete/withdraw/withdraw.tree:20]: thank you

Comment (smol-ninja) [reply tests/integration/concrete/withdraw/withdraw.tree:20]: https://github.com/sablier-labs/flow/pull/368/commits/002a055d11fafc1acd171fa7a911121ccc5739fb

**Comment by smol-ninja** [comment]
@andreivladbrg I have pushed a new commit, https://github.com/sablier-labs/flow/pull/368/commits/002a055d11fafc1acd171fa7a911121ccc5739fb, that resolved your comment about running test for all users as well as few minor changes.

Please let me know if it looks good now.

PS: I have added assertions to make sure withdraw amount satisfies conditions for each test.

**Comment by smol-ninja** [comment]
Thanks.

1. Why did you replace eve with operator where the caller is unknown? Isn't eve more appropriate to represent an unknown caller?
2. Those assert statements were left intentional to confirm the test constraints. The assert that you added in `_test_Withdraw` isn't much useful because that will always be true.

**Comment by andreivladbrg** [comment]
> Why did you replace eve with operator where the caller is unknown? Isn't eve more appropriate to represent an unknown caller?
 
as the commit message says, we use `users.eve` when a test should revert (eve from "evil" 😈)

> Those assert statements were left intentionally to verify the test constraints. The assert that you added in _test_Withdraw isn't much useful because it will always be true. 

why would using `assert` be useful and `assertTrue` not ? they do the same thing

> So I think we should bring them back as they verify the test constraints

`assertTrue` can also verify the tests constraints, but in case it reverts, foundry's version is friendlier with debugging

**Comment by smol-ninja** [comment]
> as the commit message says, we use `users.eve` when a test should revert (eve from "evil" 😈)

Ooh. I thought eve is cute 🤣

> why would using assert be useful and assertTrue not ? they do the same thing

They are same but not the conditions. In https://github.com/sablier-labs/flow/pull/368/commits/247f716f26b5788f53a7987a05edf621eb224eef, some of the asserts check for strict inequality whereas the one you added checks for `<=`. I didn't have objection with using `assertTrue` over `assert` but the conditions that it is checking.

**Comment by andreivladbrg** [comment]
> I didn't have objection with using assertTrue over assert but the conditions that it is checking

ok, i get it now, this is what i misunderstood. i thought you were talking specifically about `assert` vs `assertTrue`

to not have repetitive code what if we add boolean (`waEqTd` - withdraw amount equal total debt) to the function param, and then:

```solidity
if (waEqTd) {
    assertTrue(withdrawAmount <= flow.totalDebtOf(streamId), "amount <= total debt");
} else {
    assertTrue(withdrawAmount < flow.totalDebtOf(streamId), "amount < total debt");
}
```

wdyt?

**Comment by smol-ninja** [comment]
You will still end up with more lines of codes. We have one assertion per test, we will now end up having one bool update per test plus 5 additional lines in `_test_Withdraw`.

I think having assertions in individual test is a better approach.

- We simulate time and set things up for `_withdraw`
- We test if `withdrawAmount` follows constraints
- If yes, then we call `_withdraw`.

And not all tests have constraints related to total debt. For example, `test_WhenAmountLessThanSnapshotDebt` checks if `amount < initialSnapshotDebt`.

**Comment by andreivladbrg** [comment]
TBH, i don’t fully get why we need these test constraints verifications in the first place since (i) we are in control of the vars passed as `wa`, and (ii) we have the snapshot debt assertions after calling `_test_Withdraw`.
could you help be with this clarification first, please.

another thing is the use of flow functions for params before calling `flow.withdraw`. IMO, this isn’t the best approach for _concrete/unit tests_, where we should have exact values (defaults) for before/after

so, instead of these:

```solidity
uint256 initialSnapshotDebt = getDescaledAmount(flow.getSnapshotDebtScaled(defaultStreamId), 6);
uint128 withdrawAmount = uint128(initialSnapshotDebt) + WITHDRAW_AMOUNT_6D;
// to:
uint256 withdrawAmount = 3092e6;
```

**Comment by smol-ninja** [comment]
>  i don’t fully get why we need these test constraints verifications in the first place since

I'd say we don't need it as such because now tests are complete so we can remove them for sure. But given that each test is run for three users so having them while writing tests helped me with debugging to verify if all constraints are matched. 

Using these assertions also led me to make changes [like these](https://github.com/sablier-labs/flow/blob/247f716f26b5788f53a7987a05edf621eb224eef/tests/integration/concrete/withdraw/withdraw.t.sol#L113), where we warp time after taking snapshot to ensure there is ongoing debt as well. So, I don't have problems removing them but they will be helpful in the future if we make changes to the tests and want to be sure of the constraints.

> we should have exact values (defaults) for before/after

Whats wrong with it? Why is it not the best approach? Also, this is why assertions are useful here because its hard to keep track of all the variables in absolute numbers with three users.

Can I also know why you don't want to keep assertions?

**Comment by andreivladbrg** [comment]
good point about modifiers, will move back the assert in the individual functions

**Review by andreivladbrg — APPROVED** [review]
👍

**Comment by smol-ninja** [comment]
Great. Thank you.
